rainier-core/src/main/scala/com/stripe/rainier/sampler/MAP.scala

```diff
@@ -1,30 +1,33 @@
 package com.stripe.rainier.sampler
 
-import com.stripe.rainier.compute._
-
 final case class MAP(stepSize: Double) extends Sampler {
-  def sample(context: Context,
+  def sample(density: DensityFunction,
              warmupIterations: Int,
              iterations: Int,
              keepEvery: Int)(implicit rng: RNG): List[Array[Double]] = {
-    val values = MAP.optimize(context, warmupIterations, stepSize)
+    val values = MAP.optimize(density, warmupIterations, stepSize)
     List(values)
   }
 }
 
 object MAP {
-  def optimize(context: Context,
+  def optimize(density: DensityFunction,
                iterations: Int,
                stepSize: Double): Array[Double] = {
-    val cf =
-      context.compiler
-        .compile(context.variables, context.density :: context.gradient)
-    val initialValues = context.variables.map { v =>
-      0.0
-    }.toArray
+    val cf = (params: Array[Double]) => {
+      density.update(params)
+      density.gradientVector
+    }
+
+    val initialValues = 1
+      .to(density.nVars)
+      .map { _ =>
+        0.0
+      }
+      .toArray
     1.to(iterations).foldLeft(initialValues) {
       case (values, _) =>
-        val _ :: gradient = cf(values).toList
+        val gradient = cf(values).toList
         values.zip(gradient).map {
           case (v, g) =>
             v + (stepSize * g)
```

Comment:
would be pretty easy to replace this with zipWithIndex instead of zip(gradient), and then density.gradient(i), right? looking to get rid of gradientVector if possible

Reply:
ah, sure, yeah.